### ISSUE TITLE
Memory leak in cmark_parse_document

### ISSUE THREAD

**Issue by MatsMoll** [vapor-community/markdown#10]
I started noticing unusual memory in my Vapor 4 project and ran some analysis using instruments. Instruments showed that there was a memory leak in the cmark-gfm package, as shown below.

<img width="1679" alt="Screen Shot 2020-12-03 at 22 11 42" src="https://user-images.githubusercontent.com/4439131/101091531-91b8e400-35b8-11eb-9781-c3a964982aaf.png">

This leak was reproducible with the following code:

```swift
let strings = [
    "",
    "Rationality action belief follows logic reason. Rationality concerned expected actions results depending upon agent perceived. Obtaining useful information action performed big part rationality.",
    """
    Turing test, created computer scientist Alan Turing, commonly discussed theme computer world. paper 1950, wanted answer question: “Can machines think?”. get bottom this, created hypothetical test see machine capable mimicking human intelligence specific conditions.
    To conduct Turing test, researches set three separate terminals. One division operated computer, two operated humans. One human blindly question computer human specific subject. human discern answer came computer human. repeated multiple times, questioner could know answer came human half time, computer considered artificial intelligence.
    Read [Turing Test](https://en.wikipedia.org/wiki/Turing_test)
    """
]

do {
    for string in strings {
        _ = try markdownToHTML(string)
    }
} catch {
    print("Error: \(error.localizedDescription)")
}
```

<img width="1150" alt="Screen Shot 2020-12-03 at 22 12 14" src="https://user-images.githubusercontent.com/4439131/101092134-74384a00-35b9-11eb-9fb3-2e929410b61b.png">

<img width="1152" alt="Screen Shot 2020-12-03 at 22 12 39" src="https://user-images.githubusercontent.com/4439131/101092099-64206a80-35b9-11eb-81a5-abe6e159c3e0.png">

**Comment by vzsg** [comment]
It seems those extensions are [not properly cleaned up](https://github.com/brokenhandsio/cmark-gfm/blob/master/Sources/cmark/blocks.c#L160-L161), they are just released by the `cmark_llist_free` function, yet their internal state remains allocated and leaked.

Maybe those lines should be using `cmark_llist_free_full` instead, with `cmark_syntax_extension_free` passed in to tear them down properly 🤔.

**Comment by MatsMoll** [comment]
@vzsg  Interesting 🤔
It seems like `cmark_llist_free` is using a `NULL` pointer as the release function.
So I will try to provide a release function there and check it solves the problem you explained.

```c
void cmark_llist_free(cmark_mem *mem, cmark_llist *head) {
  cmark_llist_free_full(mem, head, NULL);
}
```

**Comment by MatsMoll** [comment]
So the following code reduces the number of leaks from 26 to 16, but I get a crash when adding the release function to the second linked list. For some reason 🤔
```c
void free_data_function(cmark_mem *mem, void *user_data) {
    mem->free(user_data);
}

void cmark_parser_free(cmark_parser *parser) {
  cmark_mem *mem = parser->mem;
  cmark_parser_dispose(parser);
  cmark_strbuf_free(&parser->curline);
  cmark_strbuf_free(&parser->linebuf);
  cmark_llist_free_full(parser->mem, parser->syntax_extensions, *free_data_function);
  cmark_llist_free_full(parser->mem, parser->inline_syntax_extensions, NULL);
  mem->free(parser);
}
```

**Comment by vzsg** [comment]
I'm not sure why you added that cmark_free_llist_user_data function, since there's already a proper cleanup function for syntax extensions.

The following eliminated the leaks for me, at least those that were identified by the memory debugger. The `inline_syntax_extensions` list seems to be a sublist of the larger `syntax_extensions`, so the deep cleaning of the larger one is enough.

```c
void cmark_parser_free(cmark_parser *parser) {
  cmark_mem *mem = parser->mem;
  cmark_parser_dispose(parser);
  cmark_strbuf_free(&parser->curline);
  cmark_strbuf_free(&parser->linebuf);
  cmark_llist_free_full(parser->mem, parser->syntax_extensions, cmark_syntax_extension_free);
  cmark_llist_free(parser->mem, parser->inline_syntax_extensions);
  mem->free(parser);
}
```


That being said, the whole library seems shaky at best: repeat the markdownToHTML call a few thousand times and it crashes all around the place. 🙁

**Comment by MatsMoll** [comment]
Ah, I did not know or find a dedicated free function for syntax extensions. But good to know that fixes the leaks!

And agree! Looks like it is not completely stable, so could possibly be a good idea to move to something like Ink?

**Comment by MatsMoll** [comment]
Or then it probably makes more sense to import Ink directly into your project instead of through this framework 😅

**Comment by 0xTim** [comment]
This is resolved in [0.7.0](https://github.com/vapor-community/markdown/releases/tag/0.7.0) which uses a new API which tidies up after itself. If I profile
```swift
import SwiftMarkdown

let string = """
    Turing test, created computer scientist Alan Turing, commonly discussed theme computer world. paper 1950, wanted answer question: “Can machines think?”. get bottom this, created hypothetical test see machine capable mimicking human intelligence specific conditions.
    To conduct Turing test, researches set three separate terminals. One division operated computer, two operated humans. One human blindly question computer human specific subject. human discern answer came computer human. repeated multiple times, questioner could know answer came human half time, computer considered artificial intelligence.
    Read [Turing Test](https://en.wikipedia.org/wiki/Turing_test)
    """
for i in 1...5000 {
    print("Iteration \(i)")
    _ = try markdownToHTML(string)
}
```

I get no leaks (or crashes).

There is a different issue related to extension support that crashes due to another bug in cmark which I'm currently looking into
